src/test.zig

```diff
@@ -785,108 +1010,99 @@ pub const TestContext = struct {
+            // Split filename into "<base_name>.<index>.<file_ext>"
+            const prev_parts = getTestFileNameParts(prev_filename);
+            const new_parts = getTestFileNameParts(filename);
+
+            // If base_name and file_ext match, these files are in the same test sequence
+            // and the new one should be the incremented version of the previous test
+            if (std.mem.eql(u8, prev_parts.base_name, new_parts.base_name) and
+                std.mem.eql(u8, prev_parts.file_ext, new_parts.file_ext))
+            {
+                // This is "foo.X.zig" followed by "foo.Y.zig". Make sure that X = Y + 1
+                if (prev_parts.test_index == null) return error.InvalidIncrementalTestIndex;
+                if (new_parts.test_index == null) return error.InvalidIncrementalTestIndex;
+                if (new_parts.test_index.? != prev_parts.test_index.? + 1) return error.InvalidIncrementalTestIndex;
+            } else {
+                // This is not the same test sequence, so the new file must be the first file
+                // in a new sequence ("*.0.zig") or an independent test file ("*.zig")
+                if (new_parts.test_index != null and new_parts.test_index.? != 0) return error.InvalidIncrementalTestIndex;
+                cases.clearRetainingCapacity();
             }
             prev_filename = filename;
 
             const max_file_size = 10 * 1024 * 1024;
             const src = try dir.readFileAllocOptions(ctx.arena, filename, max_file_size, null, 1, 0);
 
-            // The manifest is the last contiguous block of comments in the file
-            // We scan for the beginning by searching backward for the first non-empty line that does not start with "//"
-            var manifest_start: ?usize = null;
-            var manifest_end: usize = src.len;
-            if (src.len > 0) {
-                var cursor: usize = src.len - 1;
-                while (true) {
-                    // Move to beginning of line
-                    while (cursor > 0 and src[cursor - 1] != '\n') cursor -= 1;
-
-                    if (std.mem.startsWith(u8, src[cursor..], "//")) {
-                        manifest_start = cursor; // Contiguous comment line, include in manifest
-                    } else {
-                        if (manifest_start != null) break; // Encountered non-comment line, end of manifest
+            // Parse the manifest
+            var manifest = try TestManifest.parse(ctx.arena, src);
 
-                        // We ignore all-whitespace lines following the comment block, but anything else
-                        // means that there is no manifest present.
-                        if (std.mem.trim(u8, src[cursor..manifest_end], " \r\n\t").len == 0) {
-                            manifest_end = cursor;
-                        } else break; // If it's not whitespace, there is no manifest
-                    }
+            if (cases.items.len == 0) {
```

Comment:
IIUC, this skips over any backend/target/etc. entries in the manifest if the test case is part of an incremental update

Would it be better to check that these have been left unspecified, or else match the effective configuration?

Reply:
Yeah, this is on purpose. I don't want to created cases for incremental updates, rather I just want to update expected error/output conditions. Having said that, I think it'd be nice if we checked if the manifest changed significantly and flag an error if it did (I don't see any value in suddenly requesting a different target for an update - doesn't make any sense, right?). I didn't want to introduce too much new functionality from get-go into this PR hence why I didn't add the full manifest checking yet.

Also, note we only ignore the config values; the type of the update (`run` or `error`) as well as parsing of the trailing data is still performed as expected.

Reply:
One thing we could consider at a spec level though is: do we want to spell out full manifests for each incremental update test file so that we can run them as independent test cases, or not? For now, I assumed that latter, but we can easily change that in the future should we want more flexibility.

Reply:
Finally, this check `cases.items.len == 0` is equivalent to the previous `opt_case == null` check. The reason we need an array here is that we need to be able to unpack a manifest into potentially several cases that are a Cartesian product of the `backend`s and `target`s.

Reply:
For explicitly incremental updates (".0.zig", ".1.zig", etc.), I think that makes sense

It'd be nice to check that these config values are not provided in anything except the first config, so that folks are not confused when these are ignored. But I'm definitely OK with that being left as a TODO for now

The more important milestone will be when we try to run these tests in incremental mode, so that tests become "implicitly" incremental, too. I'd like to make sure we don't forget that we haven't resolved the configuration differences between test cases when we eventually try to turn that on for stage2. It's not totally clear to me what the current proposal is for handling backends, etc. when running in incremental mode

cc @mitchellh , since he may have some valuable input here

Reply:
Yep agreed. I actually left it unfinished for that very reason - so that we can discuss the implications and follow-up with a fix(es) if desired.

For incremental mode, we can run both `stage2` and `llvm` backends which we already do. `stage1` will assert there are updates upon execution.